### PR TITLE
Navigation Link block: register variations on post type / taxonomy registration

### LINKED ISSUE #53826: Navigation Link Block: Variations for post types may not contain all post types
### Description

Similar to #52569 the navigiation link block creates a variation for all post types. But the hook to check for public post types (especially those with `show_in_nav_menus` set to true) is run on `init` with a default priority of 10. 
It's safe to asume, that this will run before many plugins will have registered their custom post types, because the core init hook is added earlier and 10 is the default priority (so many developers will register it on 10 as well). 

File: https://github.com/WordPress/gutenberg/blob/16dcc04a63dd250f89f3bf75b21b6fd10adabe47/packages/block-library/src/navigation-link/index.php#L371

**Proposed solution:** 
1. Run the complete `register_block_core_navigation_link` on a priority later than 10 (eg 100)
2. Split the function and add variations for all public post types in the editor in JavaScript (performance for fetching all post types?)
3. ~Split the function and add variations add a later point via PHP~ (not possible because [server side version to register block variations is missing](https://github.com/WordPress/gutenberg/issues/47170))

All other core blocks are registered on priority 10, so I think it would be a bit odd to run the register function for only this block on a later priority. Also of course we can't be 100% sure if on priority 100 all custom post types are already registered (since developers also can use later priorities). So I guess I actually favor solution 2. Is there precedent for any solution like this?

See also discussion in #52569 

My attention was brought to this bug by [@kyleakelly@cosocial.ca](https://fosstodon.org/@kyleakelly@cosocial.ca). 

### Step-by-step reproduction instructions

**Negative test:**
1. Register a custom post type ([example code from developer.wordpress.org](https://developer.wordpress.org/plugins/post-types/registering-custom-post-types/)). 
```php
add_action('init', function () {
    register_post_type(
        'wporg_product',
        array(
            'labels'      => array(
                'name'          => __('Products', 'textdomain'),
                'singular_name' => __('Product', 'textdomain'),
                'item_link' => __('Product Link', 'textdomain')
            ),
            'public'      => true,
            'has_archive' => true,
            'show_in_rest' => true,
            'show_in_nav_menus' => true
        )
    );
}, 11);
```
**Important:** Set the `item_link` label or the block variation is called `Post Link`.
2. Go into site-editor and add a navigation block 
3. try to search for a block named like the custom taxonomy (eg "Product Link").
4. Block is not available, only the default "Post Link" and "Pages Link" will be availab.e

**Positive test:** 
1. Register a custom post type ([example code from developer.wordpress.org](https://developer.wordpress.org/plugins/post-types/registering-custom-post-types/)). 
```php
add_action('init', function () {
    register_post_type(
        'wporg_product',
        array(
            'labels'      => array(
                'name'          => __('Products', 'textdomain'),
                'singular_name' => __('Product', 'textdomain'),
                'item_link' => __('Product Link', 'textdomain')
            ),
            'public'      => true,
            'has_archive' => true,
            'show_in_rest' => true,
            'show_in_nav_menus' => true
        )
    );
}, 9);
```
**Important:** Set the `item_link` label or the block variation is called `Post Link`.
**Important:** change `add_action` **priority** to something lower than 10 (eg 9)
2. Go into site-editor and add a navigation block 
3. try to search for a block named like the custom taxonomy (eg "Product Link").
4. Block is available

### Screenshots, screen recording, code snippet

_No response_

### Environment info

- WordPress 5.2

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by gaambo**
## What?

This is a different approach to solving #53826. Instead of using a later priority for the init hook (as in #54434), this PR directly hooks into `registered_post_type` and `registered_taxonomy`. 

Closes https://github.com/WordPress/gutenberg/issues/53826

## Why?

See #53826 - server side function runs too early to make sure it gets all custom taxonomies/post types by plugins, themes etc..

## How?

Variations for all post types and taxonomies registered until `register_block_core_navigation_link` (at init priority 10) are registered directly in here. For all post types and taxonomies registered after that, actions are added to `registered_post_type` and `registered_taxonomy`, which will register variations during post type/taxonomy registration. 

We can't use the second approach for _all_ post types/taxonomies, because some of them (eg builtin ones) are registered before the navigation link block is registered, therefore variations can't yet be added. 

I'm directly adding the variation on the WP_Block_Type object from WP_Block_Type_Registry, because there's no API for registering variations in PHP (yet). 

#### Post Type
1. Register a custom post type ([example code from developer.wordpress.org](https://developer.wordpress.org/plugins/post-types/registering-custom-post-types/)).
```php
add_action('init', function () {
    register_post_type(
        'wporg_product',
        array(
            'labels'      => array(
                'name'          => __('Products', 'textdomain'),
                'singular_name' => __('Product', 'textdomain'),
                'item_link' => __('Product Link', 'textdomain')
            ),
            'public'      => true,
            'has_archive' => true,
            'show_in_rest' => true,
            'show_in_nav_menus' => true
        )
    );
}, 11);
```
Important: Set the item_link label or the block variation is called Post Link.
Important: change add_action priority to something higher than 10 (eg 11) - when using something lower it already works in trunk.
2. Go into site-editor and add a navigation block
3. try to search for a block named like the custom taxonomy (eg "Product Link").
4. Block is available

#### Taxonomy
1. Register a custom taxonomy ([example code from developer.wordpress.org](https://developer.wordpress.org/plugins/taxonomies/working-with-custom-taxonomies/#step-2-creating-a-new-plugin)) 
```php
add_action('init', function() {
	 $args   = array(
		 'labels'            => array(
		     'link_item'         => __( 'Course' ),
	         ),
		 'show_in_nav_menus' => true
	 );
	 register_taxonomy( 'course', [ 'post' ], $args );
});
```
Important: Set the item_link label or the block variation is called Post Link.
Important: change add_action priority to something higher than 10 (eg 11) - when using something lower it already works in trunk.
3. Go into site-editor and add a navigation block
4. try to search for a block named like the custom taxonomy (eg "Product Link").
5. Block is available

### Automated testing

Run `npm run test:unit:php:base -- --filter Block_Navigation_Link_Variations_Test`

### CODE REVIEW

**Comment by getdave** [comment]
@gaambo Thank you so much for this PR. I will review this week and will attempt to solicit feedback from WP Core contributors.

File: packages/block-library/src/navigation-link/index.php
Comment (getdave) [inline packages/block-library/src/navigation-link/index.php:341]: This makes me think we should expose a method on [`WP_Block_Type`](https://developer.wordpress.org/reference/classes/wp_block_type/) for this purpose. 

One for a follow up Issue I think.

File: packages/block-library/src/navigation-link/index.php
Comment (getdave) [inline packages/block-library/src/navigation-link/index.php]:
```suggestion
	// This will only handle post types and taxonomies registered until this point (init on priority 9).
	// See action hooks at the end of this function for other post types and taxonomies.
	// See https://github.com/WordPress/gutenberg/issues/53826 for details.
```

File: packages/block-library/src/navigation-link/index.php
Comment (getdave) [inline packages/block-library/src/navigation-link/index.php]:
```suggestion
	// Register actions for all post types and taxonomies that may not yet be registered.
	// This needs to happen in this function, because otherwise any post types/taxonomies
	// registered **before** this block will attempt to register on a block that does not yet exist.
```

**Review by getdave — COMMENTED** [review]
Initial review. I will also need to manually test this.

I'll wait for others' thoughts (especially @ironprogrammer), but I'm wondering whether these variation functions should be grouped under a static utility class in the same way we did for Navigation Fallbacks.

There have been previous discussions about such an approach as a way of avoiding creation of global functions that then need to be maintained indefinitely.

File: packages/block-library/src/navigation-link/index.php
Comment (gaambo) [reply packages/block-library/src/navigation-link/index.php:341]: There's already #47170 about registering variations on the server. Allthough for block styles in `register_block_style` it does not operate on the `WP_Block_Type` directly, but instead uses a `WP_Block_Styles_Registry`.

File: phpunit/blocks/block-navigation-link-variations-test.php
Comment (getdave) [inline phpunit/blocks/block-navigation-link-variations-test.php]: _All_ of the `assert*` will need to use the 3rd "message" parameter 🙏

File: phpunit/blocks/block-navigation-link-variations-test.php
Comment (getdave) [inline phpunit/blocks/block-navigation-link-variations-test.php:48]: Each test will need some of annotations listed here

https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#annotations

I'd say:

- covers
- ticket
- group (possibly)

File: phpunit/blocks/block-navigation-link-variations-test.php
Comment (getdave) [inline phpunit/blocks/block-navigation-link-variations-test.php]:
```suggestion
		$this-> assertNotEmpty( $nav_link_block->variations );
```

I think we can use this. See

https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#using-assertions

File: phpunit/blocks/block-navigation-link-variations-test.php
Comment (getdave) [inline phpunit/blocks/block-navigation-link-variations-test.php]: Nit: is the first assertion necessary?

File: phpunit/blocks/block-navigation-link-variations-test.php
Comment (getdave) [inline phpunit/blocks/block-navigation-link-variations-test.php]: Nit: Same question as above about doubling up on assertions

File: packages/block-library/src/navigation-link/index.php
Comment (getdave) [inline packages/block-library/src/navigation-link/index.php]: @ironprogrammer Could you confirm you would be happy with this approach?

File: packages/block-library/src/navigation-link/index.php
Comment (getdave) [inline packages/block-library/src/navigation-link/index.php]: Why might this be an array?

File: packages/block-library/src/navigation-link/index.php
Comment (getdave) [inline packages/block-library/src/navigation-link/index.php:331]:
```suggestion
function gutenberg_register_block_core_navigation_link_variation( $variation ) {
```

As per https://github.com/WordPress/gutenberg/blob/trunk/lib/README.md. Let's wait for confirmation before updating.

File: packages/block-library/src/navigation-link/index.php
Comment (getdave) [inline packages/block-library/src/navigation-link/index.php:407]:
```suggestion
function gutenberg_register_block_core_navigation_link_post_type_variation( $post_type, $post_type_object ) {
```

As per https://github.com/WordPress/gutenberg/blob/trunk/lib/README.md. Let's wait for confirmation before updating.

File: packages/block-library/src/navigation-link/index.php
Comment (getdave) [inline packages/block-library/src/navigation-link/index.php:423]:
```suggestion
function register_block_core_navigation_link_taxonomy_variation( $taxonomy, $object_type, $args ) {
```

@ramonjd I'm wondering whether we need to prefix these with `gutenberg_` as per docs in https://github.com/WordPress/gutenberg/blob/trunk/lib/README.md?

**Review by getdave — COMMENTED** [review]
This is look really good. Thank you.

A few things I noticed.

File: phpunit/blocks/block-navigation-link-variations-test.php
Comment (gaambo) [reply phpunit/blocks/block-navigation-link-variations-test.php]: Just checking that the key exists before accessing it in the next line. Should I rather do a
`$this->assertEquals( 'Custom Book', $variation['title'] ?? '' );`

File: phpunit/blocks/block-navigation-link-variations-test.php
Comment (gaambo) [reply phpunit/blocks/block-navigation-link-variations-test.php:48]: `covers` is present, `group` is set on the class, and regarding `ticket`: The documentation mentions trac issues. Is it okay to add a GitHub issue number here - there may be a trac issue with the same number and it may be confusing?

File: packages/block-library/src/navigation-link/index.php
Comment (gaambo) [reply packages/block-library/src/navigation-link/index.php]: That's according to the documentation of [registered_taxonomy](https://developer.wordpress.org/reference/hooks/registered_taxonomy/). I copied over the parameter descriptions from the documentation. We actually only need the third parameter (`$args`).

File: packages/block-library/src/navigation-link/index.php
Comment (ramonjd) [reply packages/block-library/src/navigation-link/index.php:423]: Thanks for the ping @getdave 

In general block files shouldn't use `gutenberg` prefixes. They're ported to WordPress Core mostly as-is via package sync.

There are some exceptions, but when `gutenberg_` functions are meant to run when the plugin is activated, hence we see that they're wrapped in `defined( 'IS_GUTENBERG_PLUGIN' )`.

I can add this to the lib docs too to avoid ambiguity. 

Cheers!

Comment (ramonjd) [reply packages/block-library/src/navigation-link/index.php:423]: > I can add this to the lib docs too to avoid ambiguity.

Done here: https://github.com/WordPress/gutenberg/pull/55402

Also it's worth noting there are block library-specific naming conventions outlined over in https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/README.md#naming-convention-for-php-functions

👍🏻

Comment (getdave) [reply packages/block-library/src/navigation-link/index.php:423]: Thank you Ramon. Not action require here now.

File: packages/block-library/src/navigation-link/index.php
Comment (getdave) [reply packages/block-library/src/navigation-link/index.php:407]: Ignore. I was incorrect.

File: packages/block-library/src/navigation-link/index.php
Comment (getdave) [reply packages/block-library/src/navigation-link/index.php:331]: Ignore I was incorrect.

File: phpunit/blocks/block-navigation-link-variations-test.php
Comment (getdave) [reply phpunit/blocks/block-navigation-link-variations-test.php]: Yeh I supposed it's worth checking prior to assertion. I was just thinking that the `assertEquals` would fail anyway if the `title` prop didn't exist.

File: phpunit/blocks/block-navigation-link-variations-test.php
Comment (getdave) [reply phpunit/blocks/block-navigation-link-variations-test.php:48]: I believe it has to be a Trac ticket. 

Thanks for so diligently covering off all the other annotations.

**Comment by draganescu** [comment]
This is a really cool solution @gaambo  - we're registering blocks to late for the built in cpt and taxonomies to reigster block variations, and too early for custom cpts and taxonomies to register variations 🤦🏻 

I wonder if another option would be to prerender a list of these variations and then register them clientside with `wp.blocks.registerBlockVariation`?

**Comment by gaambo** [comment]
@draganescu 
> I wonder if another option would be to prerender a list of these variations and then register them clientside with `wp.blocks.registerBlockVariation`?

I tried this approach for a similar issue (#52569 about post term block variations for taxonomies) in this PR https://github.com/WordPress/gutenberg/pull/52576

I see the following disadvantages: 
- There have to be two additional REST API calls (all post type objects, all taxonomy objects). Maybe we can preload all the data (like the navigation block does/did), but that can be a lot of data on large sites.
- Because of the REST API call, this only gets taxonomies/post types which have "show_in_rest" enabled (`show_in_rest` is a typical way to prevent a post type from using the block editor. A developer may want to use the classic editor for a post type, but still allow using it for navigation links, I guess).
- I don't know of the performance impacts of subscribe/unsubscribe, but could imagine it's kind of large?
- The _builtin property of taxonomies is not output via REST API. But by default the builtin taxonomies are registered first and therefore output first (and therefore added as a variation first).

Therefore, I went with the other solution and register them on the server side. We have all the information we need (because taxonomies and post types **have** to be registerd on the server, there are no js-only registration calls) and don't have to offload work to the client.

**Comment by draganescu** [comment]
Aweosme, thank you for all the details @gaambo 🙏🏻  

Yes. I was more leaning on prerendering serverside a big object with all this data and then register these variations in the editor boot up or when the block is init - no rest calls. So still serverside to get all the advantages you mention but without the three callbacks, all of them executed all the time.

Hm, now that I wrote the above could just enabling _everything_ be overwhelming? How could one opt out of not showing their tax or cpt in the nav block? Maybe since today the implementation doesn't work we have not had the chance to be overwhelmed with 40 different variations?

**Comment by gaambo** [comment]
Thank you for your input @draganescu!

> Yes. I was more leaning on prerendering serverside a big object with all this data and then register these variations in the editor boot up or when the block is init - no rest calls. So still serverside to get all the advantages you mention but without the three callbacks, all of them executed all the time.

Yes, that could work. Do you happen to know if there's any other place where such data is passed to the editor and where to put that date? I found that the site editor loads the [block categories](https://github.com/WordPress/wordpress-develop/blob/c9ff475e1fe63de1356bc6a3792e6fd836a8c587/src/wp-admin/site-editor.php#L134) and (still experimental) the [server side registered block defintions](https://github.com/WordPress/wordpress-develop/blob/c9ff475e1fe63de1356bc6a3792e6fd836a8c587/src/wp-admin/site-editor.php#L129). Which is both done via `wp_add_inline_script` to the `wp-blocks`. That code runs way after `init` so all post types are registered already.
I guess we could use that here as well. But maybe you/someone else has a better idea if that is the right way to add server-side generated data to the editor? 
 
> Hm, now that I wrote the above could just enabling _everything_ be overwhelming? How could one opt out of not showing their tax or cpt in the nav block? Maybe since today the implementation doesn't work we have not had the chance to be overwhelmed with 40 different variations?

I'm checking on the `show_in_nav_menus` flag for taxonomies/post types, which is on par with the classic navigation editor. So internal object types will not be used for variations. I think, working with this flag will keep the existing expectations of developers as well as users.

**Review by draganescu — APPROVED** [review]
The more I thought about this the more it made sense so 🤷🏻 . My one question left is why not use the `add_action` calls at line 397 after the `init` one, is there a need for them to be in the `register_block_core_navigation_link` block registration function?

One reason may be that if the reigistration doesnt go through we should have the other hooks, but the registration seems to never exit early and `register_block_core_navigation_link_variation` will exit early if the block is not registered.

When porting on core these filters would maybe all be in the same spot so it seems all the more a reason to not set them in the registration function of the init callback?

**Comment by gaambo** [comment]
Thank you, @draganescu .
> My one question left is why not use the `add_action` calls at line 397 after the `init` one, is there a need for them to be in the `register_block_core_navigation_link` block registration function?
> 
> One reason may be that if the reigistration doesnt go through we should have the other hooks, but the registration seems to never exit early and `register_block_core_navigation_link_variation` will exit early if the block is not registered.
> 
> When porting on core these filters would maybe all be in the same spot so it seems all the more a reason to not set them in the registration function of the init callback?

The reason is stated in the comment directly above the `add_action` calls in `register_block_core_navigation_link`: 
> This needs to happen in this function, because otherwise any post types/taxonomies registered **before** this block will attempt to register on a block that does not yet exist.

Meaning: Third parties could register a post type or a taxonomy before `init` priority 10. If we did the add_action calls outside `register_block_core_navigation_link`, those would be called. But there would not be a navigation block yet registered to which we could add the variation. With the current implemented method, all post types/taxonomies - the ones registered before init priority 10 and the ones after that - will have a variation generated. 

Another way would be to do the two `add_action` calls outside `register_block_core_navigation_link`, in `register_block_core_navigation_link_variation` somehow (eg static variable) store all post types/taxonomies that are registered when there is no `core/navigation` block (yet) and then in `register_block_core_navigation_link` use that static variable to generate the first variations. But imho that seems worse then doing add_action calls in `register_block_core_navigation_link`.

**Comment by draganescu** [comment]
Ugh, I think I read that five times in past reviews 🤦🏻  Sorry for asking the obvious @gaambo 

>  Third parties could register a post type or a taxonomy before init priority 10. 

Wouldn't these be available to `register_block_core_navigation_link` at init priority 10 via `get_post_types` and `get_taxonomies`?

The functions `register_block_core_navigation_link_taxonomy_variation` and `register_block_core_navigation_link_post_type_variation` exit early if the block doesn't exist.

What am I missing?

**Comment by gaambo** [comment]
Oh no, you are absolutely right. I completely missed the obvious thing🤦. 
You are right, every post type/taxonomy registered before `register_block_core_navigation_link`, will trigger the hooks but return early because the block is not registered yet. Then when registering the block they are collected. 

I changed that and tried to document that as well.

**Comment by gaambo** [comment]
@draganescu Is there anything for me to do or do we just wait for reviewers? 

Should I add e2e tests? I've searched in the repo and found that for example in e2e-tests/plugins/ there are some e2e tests which register post types and taxonomies and then - so I guess - test for their existence in the editor.
I've never written e2e tests and have only limited experience with PHPUnit tests, but be happy to try writing those, if I have some guidance on where they should go.

If this apporach gets merged, I also would suggest to use it for #52569 (missing post terms block variations), which is pretty similar to this.

**Comment by draganescu** [comment]
@gaambo I saw the tests were not green but did not pay attention that the failing one was optional. Nevertheless, I'll be merging. Would you be willing to open a core ticket and corresponding PR porting this change to core? (that was what the optional test was about, editing PHP changes requires a core porting PR).

**Comment by gaambo** [comment]
@draganescu Thank you 🥳
I couldn't find any documentation about backporting or what to do exactly (only [this](https://developer.wordpress.org/block-editor/contributors/code/release/) about the release process). I guess should just be a trac ticket with the reference to this PR/issue and a PR that puts the PHP file changes into wordpress core. Is that correct?

**Comment by gaambo** [comment]
@draganescu I've added [two more commits](https://github.com/gaambo/gutenberg/commits/try/navigation-link-block-cpt-variations-2) for making [PHPDoc comments](https://github.com/gaambo/gutenberg/commit/9983c1e3b6e6b9dda8f829b5b7251428b94f637c) better and adding unit tests for [private post types/taxonomies](https://github.com/gaambo/gutenberg/commit/38eaa3ad93839caab9487a3ccd2a80e9e76bae94) (having `show_in_nav_menus` set to false). Should I open a follow-up PR for those, or what's the best approach here?

**Comment by draganescu** [comment]
> I guess should just be a trac ticket with the reference to this PR/issue and a PR that puts the PHP file changes into wordpress core. Is that correct?

Actually, the changes go in the package update since we only updated the block. It's worth opening a ticket, referencing this PR and asking if leaving the code in the block is good enough. 

Generally, the PHP updates **of blocks** are landing in core via package update. But in this case given we add new functions and filters we may want to move these out of the block. Maybe 🤷🏻

**Comment by getdave** [comment]
✅ I updated the [PHP Sync Tracking Issue](https://github.com/WordPress/gutenberg/issues/57959) to note this PR does not require a backport.

File: packages/block-library/src/navigation-link/index.php
Comment (getdave) [inline packages/block-library/src/navigation-link/index.php:397]: I think these two lines maybe causing the breaking PHPUNit tests for the PHP package sync process for WordPress 6.5

https://github.com/WordPress/wordpress-develop/pull/5922

Comment (gaambo) [reply packages/block-library/src/navigation-link/index.php:397]: Maybe after #56952 was merged, we can just use the `variations_callback` function and don't need to hook into the `registered_post_type` and `registered_taxonomy` actions anymore. I can test locally. 

But do you have any idea as to why those hooks may cause phpunit timeouts?

Comment (gaambo) [reply packages/block-library/src/navigation-link/index.php:397]: @getdave I've drafted up a PR which uses the now in core available `get_block_type_variations` filter instead of the hooks mentioned above. This should give us performance benefits, but also hopefully remove the problems with the unit tests you mentioned. 

https://github.com/WordPress/gutenberg/pull/58389

The PR is just a draft/exploration, because it will only work when that filter is available (which is core 6.5+). So it definitely has some BC pitfalls. Happy for any feedback or if you think, that this PR should just be reverted so gutenberg can successfully be merged into core.

Comment (getdave) [reply packages/block-library/src/navigation-link/index.php:397]: > This should give us performance benefits, but also hopefully remove the problems with the unit tests you mentioned.

Lets bring that in. I'll try it manually here.


> why those hooks may cause phpunit timeouts?

PHPUnit probably registers/unregisters post types hundreds of times. That _could_ be it? Still only a guess.

> The PR is just a draft/exploration, because it will only work when that filter is available (which is core 6.5+). So it definitely has some BC pitfalls. 

These Gutenberg changes will be for 6.5 so it should be ok. The only issue is that the Plugin should run with the last x2 versions of WP (I think) so we'd need to gate it based on the availability of the new `variations_callback ` API in Core. Can we do that?

**Pull request merged**
